test/extensions/filters/http/jwt_authn/jwks_async_fetcher_test.cc

```diff
@@ -241,6 +246,186 @@ TEST_P(JwksAsyncFetcherTest, TestNetworkFailureFetchAndRefresh) {
+          "e": "AQAB"
+       }
+    ]
+ }
+ )";
+
+class JwksAsyncFetcherRetryingTest : public testing::Test,
+                                     public Logger::Loggable<Logger::Id::filter> {
+public:
+  JwksAsyncFetcherRetryingTest() : stats_(generateMockStats(context_.scope())) {}
+
+  void setupAsyncFetcher(const std::string& config_str, uint32_t num_failures = 0,
+                         uint32_t num_success = 1) {
+    TestUtility::loadFromYaml(config_str, config_);
+
+    EXPECT_TRUE(config_.has_async_fetch());
+    EXPECT_TRUE(config_.async_fetch().fast_listener());
+    EXPECT_TRUE(config_.has_retry_policy());
+
+    timer_ = new NiceMock<Event::MockTimer>();
+    expected_duration_ = JwksAsyncFetcher::getCacheDuration(config_);
+
+    retry_timer_ = new NiceMock<Event::MockTimer>();
+    num_retries_ = PROTOBUF_GET_WRAPPED_OR_DEFAULT(config_.retry_policy(), num_retries, 1);
+
+    context_.cluster_manager_.initializeThreadLocalClusters({"pubkey_cluster"});
+    request_ = std::make_unique<Http::MockAsyncClientRequest>(
+        &(context_.cluster_manager_.thread_local_cluster_.async_client_));
+
+    num_expected_failures_ = num_failures;
+    num_expected_successes_ = num_success;
+
+    if (num_expected_failures_ > 0 || num_expected_successes_ > 0) {
+
+      EXPECT_CALL(context_.cluster_manager_.thread_local_cluster_.async_client_, send_(_, _, _))
+          .WillRepeatedly(Invoke(
+              [this](Http::RequestMessagePtr&, Http::AsyncClient::Callbacks& cb,
+                     const Http::AsyncClient::RequestOptions&) -> Http::AsyncClient::Request* {
+                if (num_expected_failures_ > num_internal_failures_) {
+                  ++num_internal_failures_;
+                  cb.onFailure(*request_, Http::AsyncClient::FailureReason::Reset);
+                  return request_.get();
+
+                } else if (num_expected_successes_ > num_internal_successes_) {
+
+                  Http::ResponseMessagePtr response_message(
+                      new Http::ResponseMessageImpl(Http::ResponseHeaderMapPtr{
+                          new Http::TestResponseHeaderMapImpl{{":status", "200"}}}));
+
+                  response_message->body().add(jwks_text);
+                  ++num_internal_successes_;
+                  cb.onSuccess(*request_, std::move(response_message));
+                  return request_.get();
+                } else {
+                  ++num_unexpected_calls_;
+                  return nullptr;
+                }
+              }));
+    }
+
+    EXPECT_CALL(context_.dispatcher_, createTimer_(_))
+        .WillOnce(Invoke([this](Event::TimerCb timer_cb) {
+          refresh_cache_cb_ = timer_cb;
```

Comment:
Hmm,  I have concerns about this added test.  You are trying to test retry in AsyncFetcher.   Here it is way deep into internal implementation of two timer creation.   It will be very hard to maintain if implementation is changed. 

IMO,  we don't need retry test here.  retry is done in JwksFetcher,  should be tested in JwksFetcher, not here.

Reply:
agreed. removed.